Avoids exception when pandas.io.json.json_normalize contains items in…

dickreuter opened the pull request:
Continued in https://github.com/pandas-dev/pandas/pull/14583

---

When using pandas.io.json.json_normalize to parse a nested json and convert it to a dataframe, the meta parameter can be used to use fields as metadata for each record in resulting table. In some cases, not all items may contain all of the specified meta fields. This change will avoid throwing an error and output np.nan instead.

… meta parameter that don't always occur in every item of the list

robertwb committed:
BLD: Support Cython 0.25

closes #14496

Commit:
Avoids exception when pandas.io.json.json_normalize contains items in meta parameter that don't always occur in every item of the list

Commit:
Avoids exception when pandas.io.json.json_normalize contains items in meta parameter that don't always occur in every item of the list

jreback commented:
what exactly is this fixing? would need a test.

jreback committed:
COMPAT/TST: fix test for range testing of negative integers to neg powers

xref https://github.com/numpy/numpy/pull/8127    closes #14489

Author: Jeff Reback <jeff@reback.net>

Closes #14498 from jreback/compat and squashes the following commits:

882872e [Jeff Reback] COMPAT/TST: fix test for range testing of negative integers to neg powers

gfyoung committed:
BUG: Accept unicode quotechars again in pd.read_csv

Title is self-explanatory.  Affects Python 2.x only.  Closes #14477.

Author: gfyoung <gfyoung17@gmail.com>

Closes #14492 from gfyoung/quotechar-unicode-2.x and squashes the following commits:

ec9f59a [gfyoung] BUG: Accept unicode quotechars again in pd.read_csv

jreback committed:
BLD: fix 3.4 build for cython to 0.24.1

dickreuter commented:
{
	"Trades" : [{
			"general" : {
				"tradeid" : 100,
				"trade_version" : 1,
				"stocks" : [{

						"symbol" : "AAPL",
						"name" : "Apple",
						"price" : "0"

					}, {

						"symbol" : "GOOG",
						"name" : "Google",
						"price" : "0"

					}
				]
			},
		}, {
			"general" : {
				"tradeid" : 100,
				"stocks" : [{

						"symbol" : "AAPL",
						"name" : "Apple",
						"price" : "0"

					}, {
						"symbol" : "GOOG",
						"name" : "Google",
						"price" : "0"

					}
				]
			},
		}
	]
}

The following will fail because trade_version is only available in one of the two items in the list, so there is no way to output it if not all elements are exactly the same. With my change it will simply ignore it and output nan instead of throwing an error.

json_normalize(data=j['Trades'], record_path=[['general','stocks']], meta=[['general','tradeid'],['general','trade_version']])

jorisvandenbossche committed:
TST: simplify tests for GH14346 (#14502)

cswarth committed:
DOC: Expand on reference docs for read_json() (#14442)

jorisvandenbossche committed:
BUG: fix DatetimeIndex._maybe_cast_slice_bound for empty index (GH14354) (#14501)

gfyoung committed:
MAINT: Expand lint for *.py (#14516)

jorisvandenbossche committed:
BUG/ERR: raise correct error when sql driver is not installed (#14527)

When the driver was not installed, but sqlalchemy itself was, when passing a URI string, you got an error indicating that SQLAlchemy was not installed, instead of the driver not being installed. This was because the import error for the driver was captured as import error for sqlalchemy.

TomAugspurger commented:
@dickreuter Can you add that as a test and add a release note? See http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-to-the-code-base

dickreuter commented:
This would be the test, but unclear where I should store it. Any suggestions?

from unittest import TestCase
from pandas.io.json import json_normalize

```

class TestDecision(TestCase):
    def test_json_normalise_fix(self):
        j = {
            "Trades": [{
                "general": {
                    "tradeid": 100,
                    "trade_version": 1,
                    "stocks": [{

                        "symbol": "AAPL",
                        "name": "Apple",
                        "price": "0"

                    }, {

                        "symbol": "GOOG",
                        "name": "Google",
                        "price": "0"

                    }
                    ]
                },
            }, {
                "general": {
                    "tradeid": 100,
                    "stocks": [{

                        "symbol": "AAPL",
                        "name": "Apple",
                        "price": "0"

                    }, {
                        "symbol": "GOOG",
                        "name": "Google",
                        "price": "0"

                    }
                    ]
                },
            }
            ]
        }
        j = json_normalize(data=j['Trades'], record_path=[['general', 'stocks']],
                           meta=[['general', 'tradeid'], ['general', 'trade_version']])
        self.assertEqual(len(j), 4)

```

TomAugspurger commented:
Looks like those tests are all in https://github.com/pandas-dev/pandas/blob/master/pandas/io/tests/json/test_json_norm.py

You could add it as a test method under `TestJSONNormalize`

Commit:
Added documentation and test for issue #14505

Commit:
Added documentation and test for issue #14505

Commit:
Added documentation and test for issue #14505

dickreuter commented:
Test and documentation is now added.

parthea committed:
DOC: Simplify the gbq integration testing procedure for contributors (#14541)

tworec committed:
BUG: tseries ceil doc fix (#14543)

jreback commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -78,3 +78,4 @@ Bug Fixes
 
 
 - Bug in ``pd.pivot_table`` may raise ``TypeError`` or ``ValueError`` when ``index`` or ``columns`` is not scalar and ``values`` is not specified (:issue:`14380`)
+- Bug in ``pandas.io.json.json_normalize``When parsing a nested json and convert it to a dataframe, the meta parameter can be used to use fields as metadata for each record in resulting table. In some cases, not all items may contain all of the specified meta fields. This change will avoid throwing an error and output np.nan instead. (:issue '14505')
```
pls simplify. a user just wants to know does this issue pertain to them, and a short expl.

make the issue

``(:issue:`14505`)``

jreback commented on pandas/io/json.py:
```diff
@@ -792,7 +792,10 @@ def _recursive_extract(data, path, seen_meta, level=0):
-                        meta_val = _pull_field(obj, val[level:])
+                        try:
+                            meta_val = _pull_field(obj, val[level:])
+                        except:
```
don't use a bare except, list a specific exception ``KeyError``?

jreback commented on pandas/io/tests/json/test_json_norm.py:
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
         self.assertEqual(result, expected)
 
+    def test_json_normalise_fix(self):
+        j = {
```
add the issue number as a comment

jreback commented on pandas/io/tests/json/test_json_norm.py:
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
+
+                    }
+                    ]
+                },
```
this prob does not pass linting. make sure it does.

jreback commented on pandas/io/tests/json/test_json_norm.py:
```diff
@@ -225,6 +225,51 @@ def test_nested_flattens(self):
+        }
+        j = json_normalize(data=j['Trades'], record_path=[['general', 'stocks']],
+                           meta=[['general', 'tradeid'], ['general', 'trade_version']])
+        self.assertEqual(len(j), 4)
```
construct the expected frame and use ``assert_frame_equal``

jreback commented on pandas/io/json.py:
```diff
@@ -792,7 +792,10 @@ def _recursive_extract(data, path, seen_meta, level=0):
                     if level + 1 > len(val):
                         meta_val = seen_meta[key]
                     else:
-                        meta_val = _pull_field(obj, val[level:])
```
I think this should be a keyword, call it ``errors='raise'|'ignore'``. You are defining ``ignore``. Please leave the default as ``raise`` (which is the current behavior).

gfyoung committed:
BUG: Don't parse inline quotes in skipped lines (#14514)

Closes gh-14459.

brandonmburroughs committed:
BUG: Dataframe constructor when given dict with None value (#14392)

jorisvandenbossche committed:
Update ISSUE_TEMPLATE: be more explicit on where to paste the output of show_versions

Commit:
Added keyword errors {'raise'|'ignore}
Added documenation
Shortened what's new
Removed commas in dictionary for linting compatibility

Commit:
Added keyword errors {'raise'|'ignore}
Added documenation
Shortened what's new
Removed commas in dictionary for linting compatibility

dickreuter commented:
Added keyword errors {'raise'|'ignore}
Added documentation
Shortened what's new
Removed commas in dictionary for linting compatibility

jreback committed:
asv compat for py3

jorisvandenbossche committed:
BUG: don't close user-provided file handles in C parser (GH14418) (#14520)

jorisvandenbossche committed:
BUG: DataFrame.quantile with NaNs (GH14357) (#14536)

jorisvandenbossche committed:
PERF: casting loc to labels dtype before searchsorted (#14551)

jorisvandenbossche committed:
DEPR: add deprecation warning for com.array_equivalent (#14567)

pandas.core.common.array_equivalent was removed without deprecation warning.
This commits adds it back to the core.common namespace with deprecation warning

jorisvandenbossche committed:
DOC: rst fixes

jorisvandenbossche committed:
BUG/API: Index.append with mixed object/Categorical indices (#14545)

* BUG/API: Index.append with mixed object/Categorical indices

* Only coerce to object if the calling index is not categorical

* Add test for the df.info() case (GH14298)

jorisvandenbossche committed:
DOC: update whatsnew/release notes for 0.19.1 (#14573)

Commit:
Merge branch 'master' of https://github.com/dickreuter/pandas

# Conflicts:
#	doc/source/whatsnew/v0.19.1.txt

jreback commented:
you need to rebase on master

```
git rebase -i origin/master
```

dickreuter commented:
I did that earlier today. It now says: "This branch is 8 commits ahead of pandas-dev:master."

jreback commented:
maybe you didn't push it
it's not about conflicts

dickreuter commented:
My fork on github seems up to date with what I have locally, so I assume it has been pushed. Are there any further changes you expected me to implement that are not present?

jreback commented:
@dickreuter its impossible to see until you rebase on master. this should have just your commits

https://github.com/pandas-dev/pandas/pull/14505/commits

dickreuter commented:
I see, isn't it showing those commits of others only because I did a rebase of my fork (and then a local rebase of my local copy instead of a merge?). If that's a problem I could delete my fork and create a new one, then make the changes again and create a new pull request, unless you have a better suggestion.

jreback commented:
you prob just need something like

```
git fetch origin
git rebase -i origin/master
git push yourremote thisbranchname -f
```

dickreuter commented:
git fetch origin --> doesn't fetch anything as my local copy is in
sync with my remote origingit rebase -i origin/master --> noop - no
commits to pick

I think all my changes can be seen
here:https://github.com/pandas-dev/pandas/pull/14505/commits/984883796e0ee6e7355c2d5a746273c9551ead91

What may be confusing is that I also did an automatic reformatting
with pycharm to make the file conform with pep8. But those changes
only concern empty spaces.



On 3 November 2016 at 23:32, Jeff Reback <notifications@github.com> wrote:

> you prob just need something like
>
> git fetch origin
> git rebase -i origin/master
> git push yourremote thisbranchname -f
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/14505#issuecomment-258305113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABMrfm7iAtn0lL-LYi_wVB5pLLx7lS5Eks5q6m76gaJpZM4Khogo>
> .
>

jreback commented:
could also be called ``upstream``
```
(pandas) [Thu Nov 03 19:52:46 ~/pandas]$ git remote -v|grep origin
origin  https://github.com/pandas-dev/pandas.git (fetch)
origin  https://github.com/pandas-dev/pandas.git (push)
```

it doesn't matter if your branch is in sync with YOUR upstream, rather it needs to be in sync with pandas master (and on top of it), that's what a rebase is.

you need to rebase to remove all of the merges of master. you shouldn't do that, instead rebase.

dickreuter commented:
This seems to be the problem:
http://stackoverflow.com/questions/40413071/after-rebasing-my-github-fork-commits-from-others-are-in-my-pull-request/40413455#40413455

Will try to fix it, or if it's too complicated just delete and redo.

jorisvandenbossche commented:
@jreback Regarding your comment above (https://github.com/pandas-dev/pandas/pull/14505#issuecomment-258305113): pandas-dev repo is typically called 'upstream', and your own fork 'origin' (that's how our contributor guide also says it), so you need `git fetch upstream` and `git rebase -i upstream/master` instead of `git fetch origin` and `git rebase -i origin/master` to rebase properly.

jorisvandenbossche commented:
Follow-up in #14583
